src/fp3.rs

```diff
@@ -42,7 +42,46 @@ pub(crate) struct Fp3 {
 
 impl fmt::Debug for Fp3 {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        write!(f, "{:?} + {:?}*u + {:?}*u^2", self.a0, self.a1, self.a2)
+        let coeffs = [self.a0, self.a1, self.a2];
+
+        let format_term = |coef: Fp, degree: usize| -> String {
+            if coef == Fp::one() && degree > 0 {
+                let exp = if degree > 1 {
+                    format!("^{}", degree)
+                } else {
+                    "".to_string()
+                };
+                format!("u{}", exp)
+            } else {
+                let exp =  if degree > 0 {
+                    format!("u{}", if degree > 1 { format!("^{}", degree) } else { "".to_string() })
+                } else {
+                    "".to_string()
+                };
+                format!("{}{}", coef, exp)
+            }
+        };
+
+        let elem_rep = coeffs
+            .iter()
+            .enumerate()
+            .filter_map(|(i, &coef)| {
+                if coef == Fp::zero() {
+                    None
+                } else {
+                    Some(format_term(coef, i))
+                }
+            })
+            .collect::<Vec<String>>()
+            .join(" + ");
+
+        if elem_rep.is_empty() {
+            write!(f, "0")?; // Handle the case where all coefficients are zero
+        } else {
+            write!(f, "{}", elem_rep)?;
+        }
+
+        Ok(())
```

Comment:
I think it would be cleaner to handle the 0 case first, with
```rust
if *self == Fp3::zero() {
    return write!(f, "0");
}
```

so that this segment can be simply replaced by
```rust
write!(f, "{}", elem_rep)
```